File: 21.合并两个有序链表.py
#
# @lc app=leetcode.cn id=21 lang=python3
#
# [21] 合并两个有序链表
#
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next: ListNode = next


# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
  def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
    if l1 == None and l2 == None:
      return None
    pre_head = ListNode(-1)
    thead = pre_head
    while l1 != None or l2 != None:
      if l1 != None and l2 != None:
        if l1.val <= l2.val:
          thead.next = l1
          l1 = l1.next
        else:
          thead.next = l2
          l2 = l2.next
      elif l1 != None:
        thead.next = l1
        l1 = l1.next
      elif l2 != None:
        thead.next = l2
        l2 = l2.next
      thead = thead.next
    return pre_head.next

# @lc code=end


def init_nodelist(l: list):
  l1 = ListNode(l[0])
  tmp = l1
  for i in l[1:]:
    tmp.next = ListNode(i)
    tmp = tmp.next
  return l1


def print_nodelist(head: ListNode):
  vs = []
  while head != None:
    vs.append(str(head.val))
    head = head.next
  print('->'.join(vs))


if __name__ == "__main__":
  l1 = init_nodelist([1, 2, 4])
  l2 = init_nodelist([1, 3, 4])
  # print_nodelist(l1)
  # print_nodelist(l2)
  S = Solution()
  h = S.mergeTwoLists(l1, l2)
  print_nodelist(h)

File: 21.合并两个有序链表.cpp
/*
 * @lc app=leetcode.cn id=21 lang=cpp
 *
 * [21] 合并两个有序链表
 */
#include "commom.hpp"
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    ListNode *head = nullptr, *tmphead = nullptr, *node = nullptr;
    if (l1 == nullptr) {
      return l2;
    }
    if (l2 == nullptr) {
      return l1;
    }

    while (l1 != nullptr and l2 != nullptr) {
      if (l1->val <= l2->val) {
        node = l1;
        l1 = l1->next;
      } else {
        node = l2;
        l2 = l2->next;
      }
      if (head == nullptr) {
        head = node;
        tmphead = head;
      } else {
        tmphead->next = node;
        tmphead = tmphead->next;
      }
    }
    tmphead->next = l1 ? l1 : l2;
    return head;
  }
};
// @lc code=end

int main(int argc, char const* argv[]) {
  vector<int> a{1, 2, 4};
  vector<int> b{1, 3, 4};
  auto c = vector2list(a);
  auto d = vector2list(b);
  Solution s;
  auto head = s.mergeTwoLists(c, d);
  while (head) {
    ic(head->val);
    head = head->next;
  }

  return 0;
}
